File: notale-v2/experiments/workflow-skills-next/sample-workbench/mini/interaction-general/motif-match/tests/verify.py
#!/usr/bin/env python3
"""Browser contract checks for the motif-match mini candidate."""

from contextlib import contextmanager
import hashlib
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
import json
from pathlib import Path
from threading import Thread

from playwright.sync_api import sync_playwright


ROOT = Path(__file__).resolve().parents[1]
PAGES = ROOT / "candidate" / "pages"


class QuietHandler(SimpleHTTPRequestHandler):
    def log_message(self, *_):
        pass


@contextmanager
def serve():
    handler = lambda *args, **kwargs: QuietHandler(*args, directory=PAGES, **kwargs)
    server = ThreadingHTTPServer(("127.0.0.1", 0), handler)
    thread = Thread(target=server.serve_forever, daemon=True)
    thread.start()
    try:
        yield f"http://127.0.0.1:{server.server_port}/index.html"
    finally:
        server.shutdown()
        thread.join()


def snapshot(page):
    return page.evaluate("MotifMini.snapshot()")


def check_report():
    report = json.loads((ROOT / "report.json").read_text())
    total = 0
    for record in report["author_files"]:
        path = ROOT / record["path"]
        text = path.read_text()
        assert len(text) == record["chars"]
        assert hashlib.sha256(path.read_bytes()).hexdigest() == record["sha256"]
        total += len(text)
    assert total == report["mini_chars"] == 9781
    assert total <= 9800 and total < report["budget"] + 1
    for record in report["reused_files"]:
        path = ROOT / record["path"]
        assert hashlib.sha256(path.read_bytes()).hexdigest() == record["sha256"]


def choose(page, *notes):
    for note in notes:
        page.click(f'[data-note="{note}"]')


def open_page(browser, url, width=1600, height=900, reduced=False):
    page = browser.new_page(
        viewport={"width": width, "height": height},
        reduced_motion="reduce" if reduced else "no-preference",
    )
    errors = []
    page.on("pageerror", lambda error: errors.append(str(error)))
    page.on("console", lambda msg: errors.append(msg.text) if msg.type == "error" else None)
    page.goto(url, wait_until="networkidle")
    return page, errors


def check_viewports(browser, url):
    for width, height in ((1600, 900), (1280, 720)):
        page, errors = open_page(browser, url, width, height)
        assert page.locator(".key").count() == 15
        assert page.locator(".marker").count() == 8
        assert page.evaluate("[...document.querySelectorAll('.key')].map(x=>+x.dataset.note)") == list(range(60, 75))
        assert page.evaluate("document.documentElement.scrollWidth <= innerWidth")
        assert page.evaluate("document.documentElement.scrollHeight <= innerHeight")
        stage = page.locator("#stage").bounding_box()
        assert round(stage["width"]) == width and round(stage["height"]) == height
        assert not errors, errors
        page.close()


def check_learning_loop(browser, url):
    page, errors = open_page(browser, url)
    assert snapshot(page) == {
        "phase": "build", "round": 0, "notes": [60, None, None, None],
        "selected": 1, "clip": None, "markers": [0, 0, 0, 0],
    }

    choose(page, 72)
    page.click("#check")
    assert snapshot(page)["phase"] == "missing"
    assert "3、4" in page.locator("#message").inner_text()

    page.click('[data-action="reset"]')
    choose(page, 72, 71, 66)
    page.click("#check")
    assert snapshot(page)["phase"] == "error"
    assert page.locator(".interval.bad").count() == 1
    assert page.locator(".interval").all_inner_texts() == ["+12", "-1", "-5 偏 -1"]

    page.click('[data-index="3"]')
    choose(page, 67)
    assert snapshot(page)["phase"] == "edited"
    assert page.locator(".interval.bad").count() == 0
    assert page.locator(".interval").all_inner_texts() == ["+12", "-1", "-4"]
    page.click("#check")
    assert snapshot(page)["phase"] == "success"

    page.click("#check")
    assert snapshot(page)["notes"] == [62, None, None, None]
    assert snapshot(page)["phase"] == "transfer"
    choose(page, 74, 73, 69)
    page.click("#check")
    assert snapshot(page)["phase"] == "complete"
    assert page.locator(".interval").all_inner_texts() == ["+12", "-1", "-4"]

    page.click('[data-action="reset"]')
    assert snapshot(page)["notes"] == [60, None, None, None]
    assert snapshot(page)["phase"] == "build"
    assert not errors, errors
    page.close()


def check_rhythm_reset_and_dispose(browser, url):
    page, errors = open_page(browser, url)
    page.click('[data-clip="u"]')
    page.wait_for_timeout(900)
    assert snapshot(page)["markers"] == [0, 0, 0, 0]
    page.click('[data-clip="r"]')
    page.wait_for_timeout(850)
    assert snapshot(page)["markers"] == [True, 0, 0, 0]
    assert page.locator(".key.down").count() == 1

    page.click('[data-action="reset"]')
    page.wait_for_timeout(1100)
    assert snapshot(page)["markers"] == [0, 0, 0, 0]
    assert snapshot(page)["clip"] is None
    before = snapshot(page)
    page.evaluate("MotifMini.dispose()")
    page.click('[data-note="74"]')
    assert snapshot(page) == before
    page.reload(wait_until="networkidle")
    assert snapshot(page)["notes"] == [60, None, None, None]
    assert not errors, errors
    page.close()


def check_reduced_and_audio_fallback(browser, url):
    page, errors = open_page(browser, url, reduced=True)
    assert page.locator(".marker").first.evaluate("x=>getComputedStyle(x).transitionDuration") == "0s"
    page.click('[data-clip="r"]')
    page.wait_for_timeout(850)
    assert snapshot(page)["markers"][0] is True
    assert page.locator(".marker.on").count() == 1
    assert not errors, errors
    page.close()

    page, errors = open_page(browser, url + "?audio=off")
    choose(page, 72, 71, 67)
    page.click("#check")
    assert snapshot(page)["phase"] == "success"
    assert "声音不可用" in page.locator("#audioState").inner_text()
    assert not errors, errors
    page.close()


def main():
    check_report()
    with serve() as url, sync_playwright() as playwright:
        browser = playwright.chromium.launch(headless=True)
        check_viewports(browser, url)
        check_learning_loop(browser, url)
        check_rhythm_reset_and_dispose(browser, url)
        check_reduced_and_audio_fallback(browser, url)
        browser.close()
    print("PASS motif-match mini: viewport, model, rhythm, fallback, lifecycle")


if __name__ == "__main__":
    main()

File: experiments/notale-v2/workflow-skills-next/sample-workbench/mini/interaction-general/lawn-path/tests/verify.mjs
import assert from "node:assert/strict";
import crypto from "node:crypto";
import fs from "node:fs";
import http from "node:http";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { chromium } from "/tmp/notale-playwright/node_modules/playwright/index.mjs";

const testDir = path.dirname(fileURLToPath(import.meta.url));
const sampleDir = path.dirname(testDir);
const pagesDir = path.join(sampleDir, "candidate/pages");
const mime = { ".html": "text/html", ".js": "text/javascript", ".css": "text/css" };
const server = http.createServer((request, response) => {
  const relative = request.url === "/" ? "index.html" : request.url.slice(1);
  const target = path.join(pagesDir, relative);
  response.setHeader("Content-Type", `${mime[path.extname(target)]}; charset=utf-8`);
  response.end(fs.readFileSync(target));
});
await new Promise(resolve => server.listen(0, "127.0.0.1", resolve));
const url = `http://127.0.0.1:${server.address().port}/`;
const browser = await chromium.launch({ headless: true });
const errors = [];
const expectState = (state, expected) => {
  for (const [name, value] of Object.entries(expected)) assert.deepEqual(state[name], value);
};
const snapshot = page => page.evaluate(() => window.LawnMini.snapshot());
const clickMoves = async (page, names) => {
  for (const name of names) await page.click(`[data-key=${name}]`);
};
const optimal = ["ArrowRight", "ArrowDown", "ArrowRight", "ArrowUp", "ArrowRight"];
const suboptimal = [
  "ArrowRight", "ArrowLeft", "ArrowRight", "ArrowDown",
  "ArrowRight", "ArrowUp", "ArrowRight"
];

try {
  for (const viewport of [{ width: 1600, height: 900 }, { width: 1280, height: 720 }]) {
    const page = await browser.newPage({ viewport });
    page.on("pageerror", error => errors.push(error.message));
    await page.goto(url);
    await page.waitForFunction(() => window.LawnMini);
    expectState(await snapshot(page), { moves: 43, remaining: 5, repeats: 0, player: [6, 1] });
    const layout = await page.evaluate(() => ({
      scrollWidth: document.scrollingElement.scrollWidth,
      scrollHeight: document.scrollingElement.scrollHeight,
      width: innerWidth,
      height: innerHeight,
      canvasWidth: document.getElementById("board").width
    }));
    assert.ok(layout.scrollWidth <= layout.width && layout.scrollHeight <= layout.height);
    assert.ok(layout.canvasWidth >= 560);
    await page.close();
  }

  const page = await browser.newPage({ viewport: { width: 1600, height: 900 } });
  page.on("pageerror", error => errors.push(error.message));
  await page.goto(url);
  await page.click("#board");
  for (const name of optimal) await page.keyboard.press(name);
  expectState(await snapshot(page), { moves: 48, remaining: 0, repeats: 0, player: [6, 4] });
  const beforeRock = await snapshot(page);
  await page.keyboard.press("ArrowRight");
  assert.deepEqual(await snapshot(page), beforeRock);
  assert.match(await page.textContent("#status"), /石块/);
  await page.click("#assess");
  expectState(await snapshot(page), { phase: "result", moves: 48, repeats: 0 });
  assert.match(await page.textContent("#result"), /100%.*48 步.*重复无向边 0/s);
  await page.screenshot({ path: path.join(testDir, "optimal-1600x900.png") });

  await page.click("[data-action=reset]");
  await page.click("[data-action=reset]");
  expectState(await snapshot(page), { phase: "repair", moves: 43, remaining: 5, repeats: 0 });
  await clickMoves(page, suboptimal);
  expectState(await snapshot(page), { moves: 50, remaining: 0, repeats: 2, player: [6, 4] });
  await page.click("#assess");
  assert.match(await page.textContent("#result"), /96%.*50 步.*重复无向边 2/s);

  await page.click("[data-action=reset]");
  const beforeDispose = await snapshot(page);
  await page.evaluate(() => window.LawnMini.dispose());
  await page.click("[data-key=ArrowRight]");
  assert.deepEqual(await snapshot(page), beforeDispose);
  await page.reload();
  expectState(await snapshot(page), { moves: 43, remaining: 5, fallback: false });
  await page.close();

  const reduced = await browser.newPage({
    viewport: { width: 1280, height: 720 }, reducedMotion: "reduce"
  });
  await reduced.goto(url);
  const duration = await reduced.$eval("#mower", element => getComputedStyle(element).transitionDuration);
  assert.ok(parseFloat(duration) <= 0.01);
  await clickMoves(reduced, optimal);
  expectState(await snapshot(reduced), { moves: 48, remaining: 0 });
  await reduced.close();

  const fallbackContext = await browser.newContext({ viewport: { width: 1280, height: 720 } });
  await fallbackContext.addInitScript(() => {
    HTMLCanvasElement.prototype.getContext = () => null;
  });
  const fallbackPage = await fallbackContext.newPage();
  await fallbackPage.goto(url);
  expectState(await snapshot(fallbackPage), { fallback: true, moves: 43, remaining: 5 });
  assert.equal(await fallbackPage.isVisible("#fallback"), true);
  await clickMoves(fallbackPage, optimal);
  expectState(await snapshot(fallbackPage), { moves: 48, remaining: 0, repeats: 0 });
  assert.equal(await fallbackPage.locator(".cell.mowed").count(), 49);
  await fallbackPage.click("#assess");
  assert.match(await fallbackPage.textContent("#result"), /100%/);
  await fallbackPage.screenshot({ path: path.join(testDir, "fallback-1280x720.png") });
  await fallbackContext.close();

  const report = JSON.parse(fs.readFileSync(path.join(sampleDir, "report.json"), "utf8"));
  let total = 0;
  for (const file of report.author_files) {
    const content = fs.readFileSync(path.join(sampleDir, file.path), "utf8");
    const chars = [...content].length;
    const sha256 = crypto.createHash("sha256").update(content).digest("hex");
    assert.equal(chars, file.chars);
    assert.equal(sha256, file.sha256);
    assert.ok(content.split("\n").every(line => line.length <= 120));
    total += chars;
  }
  for (const file of report.reused_files) {
    const content = fs.readFileSync(path.join(sampleDir, file.path));
    const sha256 = crypto.createHash("sha256").update(content).digest("hex");
    assert.equal(sha256, file.sha256);
  }
  assert.equal(total, report.mini_chars);
  assert.ok(total <= 9800 && total < report.budget + 1);
  assert.deepEqual(errors, []);
  console.log("PASS author budget/checksums, 2 viewports, keyboard+dpad, invalid, optimal/suboptimal");
  console.log("PASS double reset, reduced motion, DOM fallback, resize, dispose/reload");
} finally {
  await browser.close();
  await new Promise(resolve => server.close(resolve));
}
